api/tests/unit/permissions/test_unit_permissions_calculator.py

```diff
@@ -0,0 +1,313 @@
+
+    # Then
+    assert user_permission_data.admin == expected_admin
+    assert user_permission_data.permissions == expected_permissions
+
+
+@pytest.mark.parametrize(
+    (
+        "user_permissions, user_admin, group_permissions, group_admin, "
+        "expected_permissions, expected_admin"
+    ),
+    (
+        (set(), False, set(), False, set(), False),
+        (
+            {VIEW_ENVIRONMENT},
+            False,
+            set(),
+            False,
+            {VIEW_ENVIRONMENT},
+            False,
+        ),
+        (
+            set(),
+            False,
+            {VIEW_ENVIRONMENT},
+            False,
+            {VIEW_ENVIRONMENT},
+            False,
+        ),
+        (set(), False, set(), False, set(), False),
+        (set(), True, set(), False, set(), True),
+        (set(), False, set(), True, set(), True),
+        (
+            {VIEW_ENVIRONMENT, UPDATE_FEATURE_STATE},
+            False,
+            set(),
+            False,
+            {VIEW_ENVIRONMENT, UPDATE_FEATURE_STATE},
+            False,
+        ),
+        (
+            {UPDATE_FEATURE_STATE},
+            False,
+            {VIEW_ENVIRONMENT},
+            False,
+            {VIEW_ENVIRONMENT, UPDATE_FEATURE_STATE},
+            False,
+        ),
+        (set(), True, set(), True, set(), True),
+    ),
+)
+def test_environment_permissions_calculator_get_permission_data(
+    environment,
+    organisation,
+    django_user_model,
+    user_permissions,
+    user_admin,
+    group_permissions,
+    group_admin,
+    expected_permissions,
+    expected_admin,
+):
+    # Given
+    user = django_user_model.objects.create(email="test@example.com")
+    user.add_organisation(organisation, OrganisationRole.USER)
+
+    group = UserPermissionGroup.objects.create(
+        name="Test Group", organisation=organisation
+    )
+    group.users.add(user)
+
+    user_environment_permission = UserEnvironmentPermission.objects.create(
+        user=user, environment=environment, admin=user_admin
+    )
+    group_environment_permission = (
+        UserPermissionGroupEnvironmentPermission.objects.create(
+            group=group, environment=environment, admin=group_admin
+        )
+    )
+
+    environment_permissions = {
+        pm.key: pm for pm in EnvironmentPermissionModel.objects.all()
+    }
+
+    for permission_key in user_permissions:
+        user_environment_permission.permissions.add(
+            environment_permissions[permission_key]
+        )
+
+    for permission_key in group_permissions:
+        group_environment_permission.permissions.add(
+            environment_permissions[permission_key]
+        )
+
+    # When
+    user_permission_data = get_environment_permission_data(
+        environment.id, user_id=user.id
+    )
+
+    # Then
+    assert user_permission_data.admin == expected_admin
+    assert user_permission_data.permissions == expected_permissions
+
+
+@pytest.mark.parametrize(
```

Comment:
I don't think we have a test case here to cover when a user has permissions directly and from a group?

Reply:
Done